File: my-folder/0124-binary-tree-maximum-path-sum/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int max;
    public int maxPathSum(TreeNode root) {
        max = Integer.MIN_VALUE;
        maxSum(root);
        return max;
    }
    public int maxSum(TreeNode node){
        if(node == null) return 0;
        int l = Math.max(maxSum(node.left), 0);
        int r = Math.max(maxSum(node.right), 0);

        int current = node.val + l + r;
        max = Math.max(max, current);
        return node.val + Math.max(l, r);

    }
}

File: my-folder/0124-binary-tree-maximum-path-sum/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

"""
Options:
MaxPath(l downwards)
MaxPath(r downwards)


"""
class Solution:
    best = float('-inf')
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        Solution.populate(self, root)
        return self.best

    def populate(self, root):
        if root is None:
            return 0
        
        lB = Solution.populate(self, root.left)
        rB = Solution.populate(self, root.right)

        self.best = max(self.best, max(0, lB) + root.val + max(0, rB))

        

        return max(0, root.val + max(0, lB, rB))
        

